src/java/org/apache/cassandra/service/reads/AbstractReadExecutor.java

```diff
@@ -247,60 +258,47 @@ public void maybeTryAdditionalReplicas()
 
         public SpeculatingReadExecutor(ColumnFamilyStore cfs,
                                        ReadCommand command,
                                        ReplicaLayout.ForToken replicaLayout,
                                        long queryStartNanoTime)
         {
-            super(cfs, command, replicaLayout, queryStartNanoTime);
-            this.replicaLayout = replicaLayout;
-        }
-
-        public ReplicaLayout.ForToken replicaLayout()
-        {
-            return replicaLayout;
-        }
-
-        public void executeAsync()
-        {
-            EndpointsForToken initialReplicas = replicaLayout().selected();
-
-            if (handler.blockfor < initialReplicas.size())
-            {
-                // We're hitting additional targets for read repair.  Since our "extra" replica is the least-
-                // preferred by the snitch, we do an extra data read to start with against a replica more
-                // likely to reply; better to let RR fail than the entire query.
-                makeDataRequests(initialReplicas.subList(0, 2));
-                if (initialReplicas.size() > 2)
-                    makeDigestRequests(initialReplicas.subList(2, initialReplicas.size()));
-            }
-            else
-            {
-                // not doing read repair; all replies are important, so it doesn't matter which nodes we
-                // perform data reads against vs digest.
-                makeDataRequests(initialReplicas.subList(0, 1));
-                if (initialReplicas.size() > 1)
-                    makeDigestRequests(initialReplicas.subList(1, initialReplicas.size()));
-            }
+            super(cfs, command, replicaLayout, replicaLayout.blockFor() < replicaLayout.selected().size() ? 2 : 1, queryStartNanoTime);
         }
 
         public void maybeTryAdditionalReplicas()
         {
             if (shouldSpeculateAndMaybeWait())
             {
                 //Handle speculation stats first in case the callback fires immediately
-                speculated = true;
-                replicaLayout = ReplicaLayout.forSpeculation(replicaLayout);
                 cfs.metric.speculativeRetries.inc();
-                // Could be waiting on the data, or on enough digests.
-                Replica extraReplica = replicaLayout().selected().get(replicaLayout().selected().size() - 1);
-                ReadCommand retryCommand = command;
+                speculated = true;
 
-                if (handler.resolver.isDataPresent() && extraReplica.isFull())
-                    retryCommand = command.copyAsDigestQuery();
+                ReadCommand retryCommand = command;
+                Replica extraReplica;
+                if (handler.resolver.isDataPresent())
+                {
+                    extraReplica = tryFind(replicaLayout().all(),
+                            r -> !replicaLayout().selected().contains(r)).orNull();
+                    assert extraReplica != null;
```

Comment:
tryFind, but then asserting it's not null? Doesn't all include pending? Does pounding count for speculation?

Reply:
Oh right no pending on reads